# Table of Contents for 4.5 Data Consistency Checks
    Importing libraries, set path, import dataset (products.csv and orders_wrangled.csv)
    Lesson: Find missing values in products, make dataset for null values and remove it from products
            Look for duplicates and drop them from products
            Export products as products_checked.csv
    Task:   Explore orders: run describe, head, dtypes
            Check for dtype in days_since_prior_order (just the column) and see if it can be changed to into a number: it did by sum
            Find missing values in orders: found several
            Address missing values: leave as is
            Look for duplicates: none
            Export orders as orders_checked.csv
            

## Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import os

## Importing Data

In [3]:
# Set Path
path=r'C:\Users\valev\Desktop\0325 Instacart Basket Analysis'

In [4]:
# Import Products Dataset
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import Orders.Wrangled Dataset
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## Lesson

In [6]:
# Find Missing Values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
# Create data set of nulls
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [8]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
df_prods.shape

(49693, 5)

In [10]:
#Create new df for only nonmissing records, excluding the 16 found above
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [11]:
df_prods_clean.shape

(49677, 5)

The number of records is now 16 less

In [12]:
# Look for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [13]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
# drop any duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [15]:
df_prods_clean_no_dups.shape

(49672, 5)

The number of records is now 5 less (there were 5 duplicates)

export it after task as
products_checked.csv

# Task

## 
Run the df.describe() function on your df_ords dataframe. 

In [16]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

It was already covered that the days of the week run from 0 to 6 which are seven days.   There are no negative values on any of the variables. I do not see that anything else needs to be investigated further except perhaps dropping the first 'unnamed' variable index.
The days since prior order min being zero could refer to two orders place on the same day. 

In [17]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

## Check for mixed-type data in your df_ords dataframe.

In [18]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

Eval_set is the only object but this column is not used and may be dropped.   Therefore, we are good with this check.  

In [19]:
df_ords['days_since_prior_order'].apply(type).value_counts()  #Check for types in colunn

days_since_prior_order
<class 'float'>    3421083
Name: count, dtype: int64

In [20]:
pd.to_numeric(df_ords['days_since_prior_order'], errors='coerce').isna().sum()   #Check if it can be converted into number

206209

It may be converted to number, thus, no mixed types. 

## Run a check for missing values in your df_ords dataframe

In [21]:
# Find Missing Values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in days_since_prior_order.   This could mean that there are that many first time customers.  

## Address the missing values using an appropriate method.

### I will leave the missing values as such.  It will get filled, or not, as the first time customer places another order.  

In [22]:
# Create data set of nulls --- for future reference, if needed
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [23]:
df_nan_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
11          11   2168274        2    prior             1                   2   
26          26   1374495        3    prior             1                   1   
39          39   3343014        4    prior             1                   6   
45          45   2717275        5    prior             1                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

## Run a check for duplicate values in your df_ords data

In [24]:
# Look for duplicates
df_dups_ords = df_ords[df_ords.duplicated()]

In [25]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates

In [26]:
df_ords[df_ords['order_id'].duplicated(keep=False)]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

products_checked.csv and orders_checked.csv

In [27]:
# Export df_ords
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [28]:
# Export df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))